### PR TITLE
[WIP] Implement Merkle Proof

### PR DESCRIPTION
**Pull request by junha1**

### CODE REVIEW

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+# Default ignored files
```
Comment (sgkim126) [inline .idea/.gitignore]: I suggest you add `.gitignore` file to your home directory.

File: src/proof.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+use crate::{Trie, TrieDB, TrieError};
+use ccrypto::{blake256, BLAKE_NULL_RLP};
+use cdb::HashDB;
+use primitives::H256;
+
+// Unit of proof.
+//#[derive(Clone, Eq, PartialEq, Debug, RlpEncodable, RlpDecodable)]
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct CryptoProofUnit<H, K, V> {
+    hash: H, // or root
```
Comment (sgkim126) [inline src/proof.rs]: Why do you make `H` to a generic parameter?
Do we need to consider other types except `H256`?

File: src/proof.rs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    fn make_proof(&self, key: &K) -> crate::Result<(CryptoProofUnit<H, K, V>, Box<dyn CryptoProvable<H, K, V>>)>;
+}
+
+// This is strongly bound to implementation of TrieDB. I want to specify that but rust has no HKT(?)
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct CryptoProof_MerkleTrie {
+    proof: Vec<Vec<u8>>, // Starts with the closest node to root.
+}
+
+pub struct CryptoProof_SkewedTree {
```
Comment (sgkim126) [inline src/proof.rs]: Do you have a use case for it?
I think proof for Merkle trie is enough for our use case.

File: src/proof.rs [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+	// TODO
+}
+
+
+impl<'db> CryptoStructure<H256, H256, Vec<u8>> for TrieDB<'db> {
+	fn make_proof<'k>(&self, key: &'k H256) 
+	-> crate::Result<(CryptoProofUnit<H256, H256, Vec<u8>>, Box<dyn CryptoProvable<H256, H256, Vec<u8>>>)> {
+		type Tunit = CryptoProofUnit<H256, H256, Vec<u8>>;
+		type Tunit_partial = (H256, NibbleSlice<'k>, Vec<u8>, bool);
```
Comment (sgkim126) [inline src/proof.rs]: `TunitPartial`
Please read https://doc.rust-lang.org/1.0.0/style/style/naming/README.html

File: src/proof.rs [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    proof: Vec<Vec<u8>>, // Starts with the closest node to root.
+}
+
+pub struct CryptoProof_SkewedTree {
+	// TODO
+}
+
+
+impl<'db> CryptoStructure<H256, H256, Vec<u8>> for TrieDB<'db> {
+	fn make_proof<'k>(&self, key: &'k H256)
```
Comment (sgkim126) [inline src/proof.rs]: Please change your editor configuration to use 4 spaces for indentation.
https://doc.rust-lang.org/1.0.0/style/style/whitespace.html

File: src/proof.rs [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+		type Tunit_partial = (H256, NibbleSlice<'k>, Vec<u8>, bool);
+
+		fn make_proof_upto(self_: &TrieDB<'_>, path: &NibbleSlice<'_>, hash: &H256) 
+		-> crate::Result<(Tunit_partial, Vec<Vec<u8>>)> {
+			let node_rlp = self_.db.get(&hash).ok_or_else(|| TrieError::IncompleteDatabase(*hash))?;
+
+			match Node::decoded(&node_rlp) {
+				Some(Node::Leaf(partial, value)) => {
+					if &partial == path {
+						Ok((hash, path, value, true))
```
Comment (sgkim126) [inline src/proof.rs]: I think it's better to define a structure instead of using a tuple.

**Review by sgkim126 — CHANGES_REQUESTED** [review]

File: src/proof.rs (hunk 3 above)
Comment (junha1) [reply src/proof.rs]: Yes in the light client

File: src/proof.rs (hunk 6 above)
Comment (junha1) [reply src/proof.rs]: Why?

File: src/proof.rs (hunk 5 above)
Comment (junha1) [reply src/proof.rs]: Ok but doesn't the formatter take care of that?

File: src/proof.rs (hunk 4 above)
Comment (junha1) [reply src/proof.rs]: Ok

File: src/proof.rs (hunk 2 above)
Comment (junha1) [reply src/proof.rs]: But there is no cost of such abstraction.
@majecty thinks such abstraction obstructs understanding code but I think abstraction helps to understand the structure.

File: src/proof.rs (hunk 6 above)
Comment (sgkim126) [reply src/proof.rs]: To name fields helps to read code.

Comment (junha1) [reply src/proof.rs]: Is it common to define a temporary struct that is used internally only in a single function?

Comment (sgkim126) [reply src/proof.rs]: Yes.

Comment (junha1) [reply src/proof.rs]: Ok.

File: src/proof.rs (hunk 4 above)
Comment (junha1) [reply src/proof.rs]: @sgkim126 'T' is ok?

Comment (sgkim126) [reply src/proof.rs]: I thought `tunit` is a word you want to use.
If you meant `type`, `Unit` and `UnitPartial` look better to me.

Comment (junha1) [reply src/proof.rs]: Ok thanks

File: .vscode/launch.json [hunk 7]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (sgkim126) [inline .vscode/launch.json:1]: Please add `.vscode` to your `.gitignore` file in home directory.

**Review by sgkim126 — CHANGES_REQUESTED** [review]

File: .vscode/launch.json (hunk 7 above)
Comment (junha1) [reply .vscode/launch.json:1]: Yes I did that and the new PR doesn't include change of .gitignore

File: .gitignore [hunk 8]
```diff
@@ -1,4 +1,4 @@
 /target
 **/*.rs.bk
-
+.idea/
```
Comment (sgkim126) [inline .gitignore:3]: Do not add it here. Please use your globla gitignore file.

**Comment by sgkim126** [comment]
Please organize the commits and write appropriate commit messages to help review.

**Pull request closed without merging**